### PR TITLE
feat: Support adding npm packages to the eszip

### PR DESCRIPTION
**Pull request by arnauorriols**
This is a draft implementation of the support for including npm packages in the eszip, honoring the ModuleGraph order. 

This PR also fixes the ordering of the modules within the eszip, which was DPS instead of BFS

Before landing this PR, the deno_node and deno_fs dependencies need to be removed.

Also note that this PR is based on 0.72.1 for testing convenience. Will rebase before getting ready to review

### CODE REVIEW

File: src/v2.rs [hunk 1]
```diff
@@ -1060,52 +1315,190 @@ impl EszipV2 {
...
+                  nv_ref = NpmPackageNvReference::new(PackageNvReference {
+                    nv: nv_ref.into_inner().nv,
+                    sub_path: Some(sub_path),
+                  });
+                }
+              }
+            }
+            let module = package.take_module(&nv_ref);
+            if let Some(module) = module {
+              modules.insert(
```
Comment (dsherret) [inline src/v2.rs:1445]: It looks like this only inserts the entrypoint, but the remaining modules in the package will be inserted at the end? Also, I think we need to figure out what packages this package depends on and ensure those are included otherwise they'll just be at the end of the eszip as well?

Overall, I think it would be good to talk to @lucacasonato about this PR and ask him review it because he understand the deploy side of things.

(I still feel like we can just get an npm snapshot and then order the packages' cjs/mjs/js files in the eszip based on that, which would also benefit cjs packages. Alternatively, some way to modify an eszip after the first load where the first load reports the order the modules should be ordered in and then the eszip gets modified to reflect that)

Comment (arnauorriols) [reply src/v2.rs:1445]: > It looks like this only inserts the entrypoint, but the remaining modules in the package will be inserted at the end? Also, I think we need to figure out what packages this package depends on and ensure those are included otherwise they'll just be at the end of the eszip as well?

It inserts each npm module in the `ModuleGraph`. If the module's `PackageNvReference` in the `ModuleGraph` is `npm:package@xx.xx.xx` (ie no sub_path), we resolve the entrypoint and load that. Only npm modules not in the `ModuleGraph` are at the end of the eszip. See [this test](https://github.com/denoland/eszip/pull/192/files/54bb954d1b3eaf242b95b87928094265a36117a9#diff-ff42aba765ab96f68cb209933964efa01404f135126af4e1cfe45d5e7c1fd0e9R3184-R3201) for a comprehensive visualization of the result.

> Overall, I think it would be good to talk to @lucacasonato about this PR and ask him review it because he understand the deploy side of things.

Sure! I understand this might be a bit controvertial, and I am happy to redesign it as necessary to resolve any concern. This is just the implementation I came up with that delivers an eszip ordered almost exactly as the order *all* modules are loaded by Deno when booting up, so any alternative implementation should be able to deliver the same eszip.

> I still feel like we can just get an npm snapshot and then order the packages' cjs/mjs/js files in the eszip based on that, which would also benefit cjs packages. 

That could be an evolution of what is here implemented. What I implemented is a mechanism to provide the source code of the npm packages into the eszip, as the `ModuleGraph` does not include it (and the `ModuleGraph` does not include all the npm packages that we want to include in the eszip, namely the cjs). If I'm not mistaken, what you propose is to use the npm snapshot to order the tail of cjs modules. That would be good indeed! I should do that before setting this PR ready to review

> Alternatively, some way to modify an eszip after the first load where the first load reports the order the modules should be ordered in and then the eszip gets modified to reflect that)

What would be the benefit comapred to what I propose here? Sounds a bit more convoluted.

Comment (arnauorriols) [reply src/v2.rs:1445]: @lucacasonato if you review this, keep in mind that @dsherret and I have already discussed OOB about not pulling deno_node and deno_fs dependencies into eszip. The most probably plan is to move the package subpath resolution to the client code (Deploy and others) and provide it as a property of `FromGraphNpmPackage`.

**Review by lucacasonato — CHANGES_REQUESTED** [review]
I am in favor of the general approach of this PR. I have multiple concerns though, in order of importance:

We can not depend on `deno_fs` and `deno_node`. This creates a circular dependency between denoland/deno and denoland/eszip, and it pulls v8 into many places it shouldn't be.

We now correctly order the first layer of NPM packages (direct dependencies from deno / JSR code) first. This is good. However, the next layer of dependencies (children of these NPM packages), are placed in random order at the end of the eszip. Take a project that has a `mod.ts` file that depends on `npm:react`. `npm:react` depends on `npm:loose-envify`. The module code for `npm:react` is correctly placed right after `mod.ts`. `npm:loose-envify` is however placed at the end of the eszip, in some random order. Other NPM packages, or other TS / JS files _could_ come first.

Two concerns on the entrypoint handling:
1. I think it is wrong, for example `npm:preact/hooks` has a subpath, but that subpath does not relate to a file - it is resolved to some specific file through `exports` in the `package.json`. The logic used for the main entrypoint would need to be used for all entrypoints.
2. I am not quite sure yet why we resolve the entrypoint at all. Is this because for CJS packages, we need access to the entrypoint during ESM import of the CJS package? Why would that be? I do not think we need access to the entrypoint source text because export analysis info is stored in the package metadata manifest. For ESM we need the source code during ESM loading, but we also need the source code for all files this entrypoint imports - those files are now at the end of the eszip though.

> cjs dependencies are at the end of the eszip by design, as they are only loaded when all the esm modules are loaded

I do not think this is entirely true. The _source code_ of CJS dependencies is not needed until ESM starts evaluating (so until the end of the ESM loading). However the _package manifest_ and `package.json` of a CJS package are needed right away (to determine that a file is CJS in the first place, and to determine the ESM visible exports of the CJS file).

So we still need to include CJS deps in the right place in the eszip, we can just move all the source code to the end of the eszip.

---

Overall, I think any improvement here is a good one, even if we do not cover all cases - right now all NPM packages are at the end of the eszip, so moving some data forward will be an improvement.

Ideally we could do dependency analysis inside of the ESM parts of the NPM module graph - this would tell us exactly which files we need to include for optimal ESM loading. Once we hit a CJS file, we can stop and insert the `package.json` and all meta files for that package at that spot, but only include the CJS files for that package at the end of the eszip (after all the ESM files).

**Comment by arnauorriols** [comment]
Discussed with @lucacasonato OOB. Summary:

TL;DR:
1. Postpone sub_path resolution to a follow-up iteration
2. Put imported modules at the head of the eszip tail, in import order, and then the rest of the npm packages, in the order they are provided by caller code. 
3. Fix bug in deploy that reads cjs module's source code prematurely. 
4. Plan follow-up iteration to deal with sub_path resolution and npm-esm dependencies. 

> However, the next layer of dependencies (children of these NPM packages), are placed in random order at the end of the eszip. 

The order of the modules in the tail of the eszip is the order in which packages are added to `FromGraphNpmPackages`. And so it is the caller responsibility to provide those in the proper order. 

> The logic used for the main entrypoint would need to be used for all entrypoints.

I think it is wrong, for example npm:preact/hooks has a subpath, but that subpath does not relate to a file - it is resolved to some specific file through exports in the package.json. Clarified that this is indeed necessary for all modules. However, entrypoint resolution cannot be performed from this crate (nor caller crates) without a refactor of deno_node. Therefore, this will remain unsolved. 

> I do not think we need access to the entrypoint source text because export analysis info is stored in the package metadata manifest. For ESM we need the source code during ESM loading, but we also need the source code for all files this entrypoint imports - those files are now at the end of the eszip though.

A bug in Deploy is causing the module loader to read the entrypoint of all `npm:` specifier imports, regardless of them being cjs or esm. This led me to believe the entrypoint was needed during module loading. A correct implementation would only need it if the module is esm (alongside its dependencies). We've agreed that in this iteration only meta-modules will be put in import order, and the modules will be put at the head of the tail. A follow-up iteration will deal with the esm special case.

**Comment by arnauorriols** [comment]
Pushed the agreed changes. @lucacasonato PTAL

File: src/v2.rs [hunk 2]
```diff
@@ -416,12 +421,12 @@ impl FromGraphNpmPackages {
...
   ) {
-    if !self.0.contains_key(package_nv_ref.nv()) {
+    if !self.packages.contains_key(package_nv_ref.nv()) {
       self
-        .0
+        .packages
         .insert(package_nv_ref.nv().clone(), FromGraphNpmPackage::default());
     }
-    let package = self.0.get_mut(package_nv_ref.nv()).unwrap();
+    let package = self.packages.get_mut(package_nv_ref.nv()).unwrap();
```
Comment (lucacasonato) [inline src/v2.rs:429]: My comment about using `.entry` still stands :)

File: src/v2.rs [hunk 3]
```diff
@@ -437,12 +442,12 @@ impl FromGraphNpmPackages {
...
```
Comment (lucacasonato) [inline src/v2.rs:450]: Ditto

File: src/v2.rs [hunk 4]
```diff
@@ -457,31 +462,62 @@ impl FromGraphNpmPackages {
...
```
Comment (lucacasonato) [inline src/v2.rs:470]: Ditto

File: src/v2.rs (hunk 2 above)
Comment (arnauorriols) [reply src/v2.rs:429]: oh, sure. I did this deliverately to avoid all the cloning, but not important anyway

**Review by dsherret — APPROVED** [review]
LGTM. We can improve this over time.

File: src/v2.rs [hunk 5]
```diff
@@ -280,6 +286,257 @@ pub struct FromGraphOptions<'a> {
...
+  }
+
+  pub fn new() -> Self {
+    Default::default()
+  }
+
+  pub fn add_package_with_maybe_meta(
+    &mut self,
+    package_id: PackageNv,
+    package_json: Option<FromGraphNpmModule>,
```
[the remark this answers is not in the document]
Comment (arnauorriols) [reply src/v2.rs:327]: Yes. And I think for now just adding all of them together is good enough.
